### PR TITLE
Endpoints for Virtual devices management

### PR DESCRIPTION
**Pull request by AnkRaiza**

### CODE REVIEW

File: src/main/firebase/FirebaseController.ts [hunk 1]
```diff
@@ -369,6 +369,17 @@ export class FirebaseDatabase {
...
     }
+
+    async createVirtualDevice(virtualDevice: any, userId: string) {
+        const virtualdeviceCreated = await this.db.ref()
+            .child("users")
+            .child(userId)
+            .child("virtualDevices")
+            .child(virtualDevice.id)
+            .set(virtualDevice);
+        console.log(virtualdeviceCreated);
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: remove

Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: remove

File: src/main/firebase/FirebaseController.ts [hunk 2]
```diff
@@ -369,6 +369,17 @@ export class FirebaseDatabase {
         });
         return messages;
     }
+
+    async createVirtualDevice(virtualDevice: any, userId: string) {
+        const virtualdeviceCreated = await this.db.ref()
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: you can just return await, but the await is only necessary if you want to throw the exception here, if not, just return

File: src/main/firebase/FirebaseController.ts (hunk 1 above)
Comment (ecruzado) [inline src/main/firebase/FirebaseController.ts]: remove console

File: src/main/services/v2/CreateVirtualDevice.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        if (result.error) {
+            res.status(400).send(result.error.message);
+            return res;
+        }
+        const { user_id, token, sm_api_access_token } = result.value;
+        try {
+            await saveAVSLink(user_id, token, db, sm_api_access_token);
+        } catch (err) {
+            console.error("[Services.linkAVS] failed to link AVS: ", err);
+            return res.status(500).send(UnknownError.message);
```
Comment (ecruzado) [inline src/main/services/v2/CreateVirtualDevice.ts]: looks weird to me to only just the message, if that is the case UnknownError just need to be a string

File: src/main/services/v2/DeleteVirtualDevice.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import * as Express from "express";
+import * as Returns from "../v1/Returns";
+import * as Joi from "joi";
+import * as Admin from "firebase-admin";
+import {FirebaseDatabase} from "../../firebase/FirebaseController";
+
+const UnknownError = new Error("Sorry, something went wrong, we are looking into it.");
```
Comment (ecruzado) [inline src/main/services/v2/DeleteVirtualDevice.ts]: same here

File: src/main/services/v2/CreateVirtualDevice.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+export async function saveAVSLink(dashboard_user_id: string, virtual_device_user_id: string, db: Admin.database.Database, smAPIAccessToken?: string) {
+    const fbDB = new FirebaseDatabase(db);
+    if (!smAPIAccessToken) {
+        const response: AxiosResponse = await Axios.get(`${Config.VirtualDeviceURL()}/refresh_token?user_id=${virtual_device_user_id}`);
+        smAPIAccessToken = response ? response.data.token : ""
+    }
+    const userObj = await db.ref().child("users").child(dashboard_user_id).once("value");
+    const user = userObj.val();
+    let virtualDevices = user.virtualDevices;
```
Comment (jperata) [inline src/main/services/v2/CreateVirtualDevice.ts]: const

File: src/main/services/v2/DeleteVirtualDevice.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        if (result.error) {
+            res.status(400).send(result.error.message);
+            return res;
+        }
+        try {
+            const {userId, token} = req.body;
+            const fbDB = new FirebaseDatabase(db);
+            const userObj = await db.ref().child("users").child(userId).once("value");
+            const user = userObj.val();
+            let virtualDevices = user.virtualDevices;
```
Comment (jperata) [inline src/main/services/v2/DeleteVirtualDevice.ts]: const

File: src/main/services/v2/UpdateVirtualDevice.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            return res;
+        }
+        try {
+            const {userId, token, name, status} = req.body;
+            const fbDB = new FirebaseDatabase(db);
+            const userObj = await db.ref().child("users").child(userId).once("value");
+            const user = userObj.val();
+            let virtualDevices = user.virtualDevices;
+            virtualDevices[token] = {...virtualDevices[token], name, status};
+            fbDB.linkAVS(userId, {virtualDevices});
```
Comment (ecruzado) [inline src/main/services/v2/UpdateVirtualDevice.ts]: this function is not returning anything

File: src/main/services/v2/CreateVirtualDevice.ts (hunk 5 above)
Comment (AnkRaiza) [reply src/main/services/v2/CreateVirtualDevice.ts]: im altering virtualDevices in the next line

File: src/main/services/v2/DeleteVirtualDevice.ts (hunk 6 above)
Comment (AnkRaiza) [reply src/main/services/v2/DeleteVirtualDevice.ts]: im altering virtualDevices in the next line

File: src/main/services/v2/CreateVirtualDevice.ts (hunk 5 above)
Comment (jperata) [reply src/main/services/v2/CreateVirtualDevice.ts]: adding a value in an array is not "mutating" the array, so you can use const

File: src/main/services/v2/DeleteVirtualDevice.ts (hunk 6 above)
Comment (jperata) [reply src/main/services/v2/DeleteVirtualDevice.ts]: removing a value in an array is not "mutating" the array, so you can use const

File: src/main/services/v2/CreateVirtualDevice.ts (hunk 5 above)
Comment (AnkRaiza) [reply src/main/services/v2/CreateVirtualDevice.ts]: this is an object not an array

**Pull request merged**